### PR TITLE
[Traceable FSDP2][Inductor] Re-inplace all_gather_into_tensor

### PR DESCRIPTION
**Pull request by yf225**
FSDP2 eager pre-allocates the output buffer for AllGather and the AllGather just writes into that buffer. However, under compile, by default we use out-of-place AllGather, which means in Traceable FSDP2 case we will be unnecessarily using more memory than eager. We want to re-inplace that AllGather instead.

This PR adds a post_grad pass to re-inplace all_gather_into_tensor (i.e. changing it from `all_gather_into_tensor.default` out-of-place op to `all_gather_into_tensor_out.default` out-variant op).

One thing to note is that since with this pass we are introducing a mutable op into the post_grad FX graph, we must do this pass after `reinplace_inplaceable_ops` (at which point we are okay again with having mutable ops in the graph). To facilitate this, this PR adds a `post_grad_custom_post_reinplace_pass` extension point to allow user-defined post-reinplace FX passes.

---

Test commands:
- `pytest -rA test/distributed/_composable/fsdp/test_fully_shard_compile.py::TestFullyShardCompile::test_transformer_fullgraph_backend_inductor`

---

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130786
* __->__ #129773

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @tianyu-l

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
Looks good but two questions - 

- is there a reason to not turn on this by default ?

- could we do separate config that so that this will still work with graph cache

File: torch/_inductor/config.py [hunk 1]
```diff
@@ -109,10 +110,18 @@ def fx_graph_remote_cache_default():
...
+# def my_custom_post_reinplace_pass(graph: torch.fx.graph.Graph):
+#     # my custom graph optimization pass
+#     ...
+#
 # torch._inductor.config.post_grad_custom_pre_pass = my_custom_pre_pass
 # torch._inductor.config.post_grad_custom_post_pass = my_custom_post_pass
+# torch._inductor.config.post_grad_custom_post_reinplace_pass = my_custom_post_reinplace_pass
 post_grad_custom_pre_pass: Optional[Callable[[torch.fx.graph.Graph], None]] = None
 post_grad_custom_post_pass: Optional[Callable[[torch.fx.graph.Graph], None]] = None
+post_grad_custom_post_reinplace_pass: Optional[
```
[the remark this answers is not in the document]
Comment (yf225) [reply torch/_inductor/config.py:122]: yes I think we can turn it on by default

Wonder does the pickle limitation mean we can't do graph caching if we pass any custom pass into `post_grad_custom_post_pass` (FX IR) or `reorder_for_compute_comm_overlap_passes` (Inductor IR)? or is the limitation specific to post-grad FX IR?

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -6792,7 +6792,14 @@ def create_inplace(
         packed.cpp_kernel_name = cpp_kernel_name
         packed.python_kernel_name = python_kernel_name
 
-        mark_node_as_mutating(packed, *pytree.tree_leaves(inputs))
+        inps = pytree.tree_leaves(inputs)
+        mark_node_as_mutating(packed, *inps)
+        # For inplace collective ops, the input is guaranteed to be alias of the returned value of op.
+        packed.alias_names.extend([inp.get_name() for inp in inps])
```
[the remark this answers is not in the document]
Comment (yf225) [reply torch/_inductor/ir.py:6798]: yes it's new - I think the existing code is missing this important line

**Review by eellison — APPROVED** [review]

**Pull request closed without merging**